### PR TITLE
Fix delete bug in repeating groups

### PR DESCRIPTION
**Pull request by bjosttveit**
Co-authored-by: Magnus Revheim Martinsen <mrmartinsen.96@gmail.com>

## Description

Since updating the flattenObject function, empty repeating groups get added to the formData object like this: `'Innflytter.TidligereBosteder': ''`. If there are elements in the repeating groups this does not happen. This is because `[].toString() = ''`. This then causes a problem inside `getRepeatingGroups`: since:
```typescript
if (groupFormData && groupFormData.length > 0) { // Passes this check
  const maxIndex = getMaxIndexInKeys(groupFormData);
  if (maxIndex !== -1) { // Fails this check
```
Results in the group not being added to `uiConfig.repeatingGroups`.

Which again is the cause of the bug @Magnusrm described about deleting elements.

### CODE REVIEW

File: src/altinn-app-frontend/src/utils/formLayout.ts [hunk 1]
```diff
@@ -106,7 +106,7 @@ export function getRepeatingGroups(formLayout: ILayout, formData: any) {
   });
 
   // filter away groups that should be rendered as child groups
-  const filteredGroups = formLayout.filter((group) => childGroups.indexOf(group.id) === -1);
+  const filteredGroups = groups.filter((group) => childGroups.indexOf(group.id) === -1);
```
Comment (bjosttveit) [inline src/altinn-app-frontend/src/utils/formLayout.ts:109]: Just noticed this as well. Is it supposed to check all elements in the formlayout and only filter away children of groups? Or is it sufficient to start with only the groups before filtering. It is called filteredGroups after all 🤷

**Review by olemartinorg — APPROVED** [review]

File: src/altinn-app-frontend/src/utils/databindings.ts [hunk 2]
```diff
@@ -171,7 +171,7 @@ export function flattenObject(data: any): any {
   for (const key of Object.keys(flat)) {
     if (flat[key] === null) {
       delete flat[key];
-    } else if (flat[key] === '' && key.indexOf('.') > 0) {
+    } else if (flat[key].toString() === '' && key.indexOf('.') > 0) {
```
[the remark this answers is not in the document]
Comment (bjosttveit) [reply src/altinn-app-frontend/src/utils/databindings.ts:174]: Btw, should i maybe add an explicit case for this by adding something like this instead?
```typescript
} else if (flat[key] === []) {
  delete flat[key];
}
```
Im not sure if the restriction of deeply nested (`key.indexOf('.') > 0`) is quite right here?

[the remark this answers is not in the document]
Comment (olemartinorg) [reply src/altinn-app-frontend/src/utils/databindings.ts:174]: Test against the old implementation, be as specific as possible - that's my suggestion! :+1: Also test what happens with empty objects, not just empty arrays. And test again if empty objects and empty arrays behave differently in deeper structures (that's the case with empty strings, which is why that restriction is in place - see the relevant test case for it).
